src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala

```diff
@@ -210,31 +215,62 @@ private[internal] trait TypeConstraints {
           }
         })
         if (!cyclic) {
+          def propagateBounds(tparam2: Symbol): Unit = {
+            def addBound(bound: Type) =
+              if(bound != NoType) {
+                debuglog(s"$tvar ${if(up) "addHiBound" else "addLoBound"} $bound.instantiateTypeParams($tparams, $tvars)")
+                if(up) tvar addHiBound bound
+                else tvar addLoBound bound
+              }
+            val bound0 = if(up) tparam2.info.bounds.lo.dealias else tparam2.info.bounds.hi.dealias
+
+            if(false) {
+              bound0 match {
+                case TypeRef(_, `tparam`, _) => addBound(tparam2.tpeHK.instantiateTypeParams(tparams, tvars))
+                case _ =>
+              }
+            } else {
+              // MS: should this be the implementation of typeConstructor on TypeVar?
+              def typeConstructor(tp: Type): Type = tp match {
+                case tv: TypeVar => tv.inst.typeConstructor
+                case _ => tp.typeConstructor
+              }
+              def nthArg(n: Int, tp: Type): Type = tp match {
+                case TypeRef(_, _, args) => if(n < args.size) args(n) else NoType
+                case tv: TypeVar if tv.instWithinBounds => nthArg(n, tv.inst)
+                case _ => NoType
+              }
+
+              // Try harder to propagate information we can discover from this type variables occurence in
+              // bounds on other type variable. Also avoid some ill-kinding issues.
+              // Open issue: should we track variance and and chose whether to add a high or low bound accordingly?
+              def propagate(bound0: Type, instantiate: => Type): Unit = bound0 match {
+                case TypeRef(_, `tparam`, _) if !tvar.isHigherKinded && !bound0.isHigherKinded =>
+                  addBound(instantiate)
+                case TypeRef(_, `tparam`, _) if tvar.isHigherKinded && bound0.isHigherKinded && tparam.typeParams.size == bound0.typeParams.size =>
```

Comment:
I'm confused. When can this happen? If `bound0.isHigherKinded` shouldn't it be a `PolyType` and not a `TypeRef`? Also checking the size of type parameters might not be enough. Basically we care only if the bound is `F[x, y, z[_]] <: G[x, y, z]`, but it could be `F[x, y, z[_]] <: G[z, x, y]` or `F[x, y, z[_]] <: G[x, y, List]`.

Reply:
I can't remember the scenario now. It's possible that it's something that came up with incorrect code.